### ISSUE TITLE
Add manifest option `clock_skew` to skew the clocks of certain nodes

### ISSUE THREAD

**Issue by sergio-mena** [cometbft/cometbft#2453]
In order to test PBTS in a testnet with unsynchronized clocks we need to add clock-skew functionality
* [x] Investigate that is the best (=most surgical) way to introduce clock skew to Docker containers (for DO it's trivial: set the host's system time)
* Implementation
  * [x] Add skew functionality on `cmttime`
    * #2454
  * [x] Review usage of `cmttime` in key packages
    * #2454
  * [x] Extend `e2e` to use the skew functionality
    * #2454
  * [x] Extend the `e2e` manifest generator to create (working) testnets with skewed nodes
    * #2488

**Comment by sergio-mena** [comment]
Regarding the first point in the issue's description, after some research, these are the findings:

* as container processes run on the host's kernel, docker does not offer an off-the-shelf way to change the time that a process running inside a container reads
* changing the host's time is gross, as it would affect all process and container in the system, and defeats the purpose as all containers would still read the same (skewed) clock
* there are other solutions that intercept system calls (e.g., `faketime`), which use some trick based on `LD_PRELOAD`. However due to the way golang (statically) links the binaries and also how it uses the _vdso_ directly, these surgical solutions won't work for us
* the only interception solution we found that would work for us it [this one](https://chaos-mesh.org/blog/simulating-clock-skew-in-k8s-without-affecting-other-containers-on-node/) but it is quite complex, looks brittle, and we would been to bring over (or depend on) large packages to manipulate OS processes

In conclusion, the approach to follow is the following:

*  augment `cmttime` package with a way to introduce skew, all behind a build tag (so that code is not present in production)
* review the parts of the code (consensus, types, state) that deals with time related to blocks, votes, and proposals. Make sure they use `cmttime` and not `time`. Also make use `time.Until` and `time.Since` is not used in that context
* introduce an `e2e` manifest knob, `clock_skew`, per node, that activates the new skew functionality in `cmttime`

**Comment by cason** [comment]
> Also make use time.Until and time.Since is not used in that context

What do you mean here? Do we need to mock them too? I don't think it is a bad idea.

**Comment by sergio-mena** [comment]
@cason Sorry, "Also make use" should read "Also make **sure**". I just edited the description.

No need to mock `time.Until` or `time.Since`:

* only the `time.Until` and `time.Since` that take a `cmttime.Now()` value need to be nuked, as they do `time.Now()` internally, the rest of them can be left alone for the time being
* When I needed to nuke `time.Since(t)`, I replaced it by `cmttime.Now().Sub(t)`
* When I needed to nuke `time.Until(t)`, I replaced it by `t.Sub.(cmttime.Now())`

**Comment by sergio-mena** [comment]
> Why not offering this methods as cmttime.Until and cmttime.Since?

Indeed

**Comment by sergio-mena** [comment]
Closed by #2454 and #2488
